### ISSUE TITLE
cmd/go: errors while testing go modules 

### ISSUE THREAD

**Issue by SuchithArodi** [golang/go#28336]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
$ go version
go version go1.11 darwin/amd64

### Does this issue reproduce with the latest release?
Yes 

### What operating system and processor architecture are you using (`go env`)?
$ go env 
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/sarodi/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/sarodi/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/Cellar/go/1.11/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.11/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/sarodi/temp/stream-service/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build557600271=/tmp/go-build -gno-record-gcc-switches -fno-common"

### What did you do?
Trying the new go modules. 
Currently, the repo uses the dep and I want to move from dep to go modules. 

Steps followed to move to go modules from dep (As described in https://github.com/golang/go/wiki/Modules):

1) In the project root directory create the mod file go.mod:  go mod init <path> (This will read from the Gopkg.lock file for all the dependency information)
2) Build the module. This will create a go.sum file in the root directory: go build ./...
3) Test the module: go test ./... 

It builds fine, but when I do go test ./... I see this error. 
```
$ go test ./...
# /Users/sarodi/temp/stream-service/pkg/txhandler.test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b391/_testmain.go:12:2: import path cannot be absolute path
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b391/_testmain.go:20:24: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b391/_testmain.go:22:21: undefined: _test
# /Users/sarodi/temp/stream-service/pkg/util.test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b394/_testmain.go:12:2: import path cannot be absolute path
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b394/_testmain.go:20:21: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b394/_testmain.go:22:20: undefined: _test
# /Users/sarodi/temp/stream-service/pkg/controller.test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:12:2: import path cannot be absolute path
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:20:19: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:22:17: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:24:17: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:26:16: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:28:25: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:30:18: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:32:25: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:34:22: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:36:18: undefined: _test
/var/folders/5g/yq5b2dfd7bgf424t0ff3dcmc0000gn/T/go-build102201698/b372/_testmain.go:36:18: too many errors
?   	/Users/sarodi/temp/stream-service/cmd/streamservice	[no test files]
FAIL	/Users/sarodi/temp/stream-service/pkg/controller [build failed]
?   	/Users/sarodi/temp/stream-service/pkg/mocks	[no test files]
FAIL	/Users/sarodi/temp/stream-service/pkg/txhandler [build failed]
FAIL	/Users/sarodi/temp/stream-service/pkg/util [build failed]
```

when I do 'go test all' I see this error: 

```
$ go test all
go: finding github.com/pierrec/lz4/v2/internal/xxh32 latest
go: finding github.com/onsi/ginkgo/extensions/table latest
go: finding github.com/Shopify/toxiproxy/client latest
go: finding github.com/pierrec/lz4/v2/internal latest
go: downloading github.com/pierrec/lz4/v2 v2.0.7
go: finding github.com/onsi/ginkgo/extensions latest
can't load package: package github.com/pierrec/lz4/v2/internal/xxh32: unknown import path "github.com/pierrec/lz4/v2/internal/xxh32": cannot find module providing package github.com/pierrec/lz4/v2/internal/xxh32
```

### What did you expect to see?
I expected the transition from dep to go mod with ease. 

### What did you see instead?
Errors !

**Comment by bcmills** [comment]
The first error message says `import path cannot be absolute path`. Does your code have an `import` statement that uses an absolute path?

**Comment by bcmills** [comment]
The second error message seems to be accurate. It says `cannot find module providing package github.com/pierrec/lz4/v2/internal/xxh32`. If we look at the latest `v2` tag for that repository ([`v2.0.7`](https://github.com/pierrec/lz4/tree/v2.0.7)), we find that it [declares its module](https://github.com/pierrec/lz4/blob/635575b42742856941dbc767b44905bb9ba083f6/go.mod#L1) to be `github.com/pierrec/lz4` (note the lack of `/v2`).

See also https://github.com/pierrec/lz4/issues/33.

**Comment by SuchithArodi** [comment]
> The first error message says `import path cannot be absolute path`. Does your code have an `import` statement that uses an absolute path?

I am not aware of any absolute paths in the import statements other than "github.com/...". Basically, all the import paths I see starts with "github.com/..." or a similar github private repo. I have "sync", "time", "container/list" but these are golang packages. 

May be it is complaining about the absolute path in some dependency which I use ?

**Comment by SuchithArodi** [comment]
So, while initiating the mod I was giving an absolute path for the module like 'go mod init /user/temp' this was the cause of the first error! 

Now, when I change the module path to something else like 'go mod init user/temp' it works fine. 
Can you please direct me to some resources where I can learn how exactly the go modules works. 

But, if I do 'go test all' I still see this error 
```
$ go test all
go: finding github.com/pierrec/lz4/v2/internal/xxh32 latest
go: finding github.com/Shopify/toxiproxy/client latest
go: finding github.com/onsi/ginkgo/extensions/table latest
go: finding github.com/pierrec/lz4/v2/internal latest
go: downloading github.com/pierrec/lz4/v2 v2.0.7
go: finding github.com/onsi/ginkgo/extensions latest
can't load package: package github.com/pierrec/lz4/v2/internal/xxh32: unknown import path "github.com/pierrec/lz4/v2/internal/xxh32": cannot find module providing package github.com/pierrec/lz4/v2/internal/xxh32
```

**Comment by agnivade** [comment]
The `v2` change has been reverted for that repo - https://github.com/pierrec/lz4/commit/bb6bfd13c6a262f1943c0446eb25b7f54c1fb9a2. You can remove that. The repo README is out of date.

Please see this https://golang.org/cmd/go/#hdr-Module_maintenance for more info.

**Comment by bcmills** [comment]
It appears that all of the errors are now accounted for. Is there anything more to do?

**Comment by rsc** [comment]
Closing this issue as fixed in part and covered by #28389 for the rest.
